### PR TITLE
Finish implementing use and item count caching of remote inventories for gadgets.

### LINKED ISSUE #137: [Feature request] Allow binding of gadgets to an inventory
As it is right now, there is no way to build large amounts of blocks repeatedly if one doesn't have the space in their inventory to do so. Would it be possible to make the gadgets 'bindable' to inventories like chests so that the gadgets pull from those inventories?

### PR DESCRIPTION
**Pull request by Phylogeny**
~~Dimension specification (I left some TODOs) and copy/paste tool support (can't set remote inventories for them) is not fully implemented, but I'll leave that to you Dire.~~
Everything is implemented now. Resolves #137.

### CODE REVIEW

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java [hunk 1]
```diff
@@ -113,16 +118,32 @@ public static boolean useItem(ItemStack itemStack, EntityPlayer player, int coun
     }
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java]: Would a general interface make sense, which would then be implemented by some kind of wrapper for the world (save?) and the RemoteInventoryCache?

File: src/main/java/com/direwolf20/buildinggadgets/client/RemoteInventoryCache.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import net.minecraft.util.math.BlockPos;
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
+
+@SideOnly(Side.CLIENT)
+public class RemoteInventoryCache {
+    private boolean isCopyPaste;
+    private Pair<Integer, BlockPos> locCached;
+    private Multiset<UniqueItem> cache;
+    private Stopwatch timer;
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/client/RemoteInventoryCache.java]: Why not use a Guava Cache?

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java [hunk 3]
```diff
@@ -113,16 +118,32 @@ public static boolean useItem(ItemStack itemStack, EntityPlayer player, int coun
...
+                count += cacheInv.count(new UniqueItem(itemStack.getItem(), itemStack.getMetadata()));
+        } else if (world != null) {
+            BlockPos tePos = GadgetUtils.getBoundTE(tool, world);
+            if (tePos != null) {
+                TileEntity te = world.getTileEntity(tePos);
+                if (te != null) {
+                    IItemHandler cap = te.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, null);
+                    count += countInContainer(cap, itemStack.getItem(), itemStack.getMetadata());
+                }
+            }
```
Comment (MajorTuvok) [inline src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java]: The interface approach would make this a lot nicer...

File: src/main/java/com/direwolf20/buildinggadgets/client/RemoteInventoryCache.java (hunk 2 above)
Comment (Phylogeny) [reply src/main/java/com/direwolf20/buildinggadgets/client/RemoteInventoryCache.java]: The thing with Guava caches is that you need to be able to statically compute the values from only the keys, and this is done automatically when attempting to get a cached value that doesn't exist. In this case, the values are computed in a packet on the server from both the keys and the packet's context, then sent to the client.

Just to see, I implemented it with a Guava cache anyway, and I found that it takes about the same amount of code as the current Multiset/Stopwatch implementation, but without the nice Multiset functionality (meaning more code in InventoryManipulation). Ultimately, I think the current way is better.

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java (hunk 1 above)
Comment (Phylogeny) [reply src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java]: Implemented.

File: src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java (hunk 3 above)
Comment (Phylogeny) [reply src/main/java/com/direwolf20/buildinggadgets/common/tools/InventoryManipulation.java]: Implemented.

**Comment by Phylogeny** [comment]
webhook test

**Comment by Phylogeny** [comment]
webhook pr test 2

**Comment by Phylogeny** [comment]
webhook pr comment test 3

**Pull request merged**